### PR TITLE
Get release-workflow running

### LINKED ISSUE #36: Release strategy
How do we want to release a new version?

- who should release? -> human vs GitHub Actions?
- when to release?
- how to prepare a release? -> one `release` branch with tags or one branch for each major version? like `release-4.x`, `release-5.x`
- do we want additional tests that run in GitHub actions when merging to a release branch? GH Actions is so flexible I could even clone another repo and install the intended release package into it and run tests, so we can test adapter-compliance and even the example repo if we want. There are no limits :-)
- who is responsible for release management and the deployment keys, npm repo etc.?
- do we also want to release to GitHub packages as a redundancy factor?
- what else....?
- do we want to hold these information in a `MAINTAINERS.md? ?

### LINKED ISSUE #116: Don't run CI on draft pull requests for releases
To prevent CI from failing (and spamming our inbox) on draft pull requests (which can fail as they are reflecting work in progress) we should disable lint/tests etc. when the PR is in draft state.

According to [this discussion](https://github.community/t/dont-run-actions-on-draft-pull-requests/16817) it is solved like so:

```yml
on:
  pull_request:
    types: [review_requested, ready_for_review]
```

### PR DESCRIPTION
**Pull request by jankapunkt**
## Summary

This PR is for testing the release workflow in action until it runs through.
No publish will be done, only dry-runs

The release-tests and code-ql workflow only runs, if the PR is non-draft

## Linked issue(s)

#116 #36 

## Involved parts of the project

CI

## Added tests?

they are the tests

## OAuth2 standard

not releated

## Reproduction

not releated

### CODE REVIEW

**Review by jorenvandeweyer — DISMISSED** [review]

**Comment by jankapunkt** [comment]
sorry for the email spam on failed workflows but the paths in github actions are really a tricky one....

**Comment by jankapunkt** [comment]
@jorenvandeweyer  I now forked the express adapter to our org and updated it so we can have integration tests with new releases and see if the express adapter breaks (which it does not right now). Was a big hassle but now it's working. Can you please take a short look at it?

**Review by jorenvandeweyer — DISMISSED** [review]

**Comment by jankapunkt** [comment]
Okay I commented out the `release.yml` for now and will make a manual release, once all merged into 4.2.0 but until then we need first merge this one finally

**Comment by jankapunkt** [comment]
@Uzlopak @jorenvandeweyer @HappyZombies I need at least one reviewer in order to merge this.

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -14,6 +14,7 @@ name: "CodeQL Semantic Analysis"
 on:
   push: # all pushes
   pull_request: # all PR
+    types: [review_requested, ready_for_review] # only non-draft PR
```
Comment (Uzlopak) [inline .github/workflows/codeql-analysis.yml:17]: Interesting. Didn't know this..will probably add this to my projects.

File: .github/workflows/tests-release.yml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  # STEP 2 - basic unit tests
+
+  # This is the standard unit tests as we do in the basic tests for every PR
+  unittest:
+    name: Basic unit tests
+    runs-on: ubuntu-latest
+    needs: [audit]
+    strategy:
+      matrix:
+        node: [12, 14, 16]
```
Comment (Uzlopak) [inline .github/workflows/tests-release.yml:41]: Please add node 17. We got a regression in mongoose few weeks ago, could have been detected earlier if we would have tested also for node17. Better to also add chec for node17 and later node 18 and 19, while dropping node 12 and 17.

File: .github/workflows/tests-release.yml [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        restore-keys: |
+          ${{ runner.os }}-node-${{ matrix.node }}
+
+    # for this workflow we also require npm audit to pass
+    - run: npm ci
+    - run: npm run test:coverage
+
+    # with the following action we enforce PRs to have a high coverage
+    # and ensure, changes are tested well enough so that coverage won't fail
+    - name: check coverage
```
Comment (Uzlopak) [inline .github/workflows/tests-release.yml:65]: If we use NYC we can set minimum code coverage and then it already throws an error in the step above.

This is how I do it in a project of mine

https://github.com/cthulhu-node/benchmark.js

File: .github/workflows/tests-release.yml [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  # Since our release may affect several packages that depend on it we need to
+  # cover the closest ones, like adapters and examples.
+
+  integrationtests:
+    name: Extended integration tests
+    runs-on: ubuntu-latest
+    needs: [unittest]
+    strategy:
+      matrix:
+        node: [12, 14] # TODO get running for node 16
```
Comment (Uzlopak) [inline .github/workflows/tests-release.yml:82]: Create follow up issue after merge

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -61,15 +38,10 @@ jobs:
...
-    #   uses: devmasx/coverage-check-action@v1.2.0
-    #   with:
-    #     type: lcov
-    #     result_path: coverage/lcov.info
-    #     min_coverage: 90
-    #     token: ${{github.token}}
-    
+    # with the following action we enforce PRs to have a high coverage
+    # and ensure, changes are tested well enough so that coverage won't fail
+    - name: check coverage
```
Comment (Uzlopak) [inline .github/workflows/tests.yml:43]: The same as above.

**Review by jorenvandeweyer — APPROVED** [review]

**Review by Uzlopak — DISMISSED** [review]
We don't have linting check? Maybe should be added in a separate PR?

File: .github/workflows/tests-release.yml (hunk 2 above)
Comment (Uzlopak) [reply .github/workflows/tests-release.yml:41]: It is not a must have...

**Review by Uzlopak — COMMENTED** [review]
Hey, I don't want to be the blocker. I change my vote to comment. :)

**Comment by jankapunkt** [comment]
@Uzlopak let's open a new Issue since this can be implemented independently from the next release

**Pull request merged**